flow/project.py

```diff
@@ -2420,52 +2693,123 @@ def _register_operations(self):
             if name in self._operations:
                 raise ValueError(
                     "Repeat definition of operation with name '{}'.".format(name))
 
-            # Extract pre/post conditions and directives from function:
+            # Extract pre/post conditions from function:
             params = {key: getattr(func, '_flow_{}'.format(key), None)
-                      for key in ('pre', 'post', 'directives')}
+                      for key in ('pre', 'post')}
 
             # Construct FlowOperation:
             if getattr(func, '_flow_cmd', False):
-                self._operations[name] = FlowOperation(cmd=func, **params)
+                self._operations[name] = FlowCmdOperation(cmd=func, **params)
             else:
-                self._operations[name] = FlowOperation(
-                    cmd=_guess_cmd(func, name, **params), **params)
+                self._operations[name] = FlowOperation(name=name, **params)
                 self._operation_functions[name] = func
 
+    @classmethod
+    def make_group(cls, name, run_cmd=None, exec_cmd=None, directives=None, options=None):
+        # Gets the relative filepath of the function caller
+        cls._GROUPS.append({'name': name, 'run_cmd': run_cmd,
+                            'exec_cmd': exec_cmd, 'directives': directives,
+                            'options': options})
+
+        # Create decorator that adds a _flow_group label to operations.
+        def add_to_group(func):
+            if hasattr(func, '_flow_group'):
+                if add_to_group.group_name in func._flow_group:
+                    raise ValueError("Can't specify the same operation "
+                                     "for a group twice.")
+                func._flow_group.append(add_to_group.group_name)
+            else:
+                func._flow_group = [add_to_group.group_name]
+            return func
+        # Specify group name
+        add_to_group.group_name = name
+        return add_to_group
+
+    def _register_groups(self):
+        "Register all groups and add the correct operations to each."
+        groups = []
+        # Gather all groups from class and parent classes.
+        for cls in type(self).__mro__:
+            groups.extend(getattr(cls, '_GROUPS', []))
+
+        # Initialize all groups without operations
+        for group in groups:
+            # Raise error for duplicate groups
+            if group['name'] in self._groups:
+                raise ValueError(
+                    "Repeat definition of group with name " +
+                    "'{}'.".format(group['name']))
+
+            # Getting the path of the FlowProject is difficult. Using the line
+            # of code below can lead to situations where the flow/project.py
+            # module is choosen
+            # path = inspect.getsourcefile(inspect.getmodule(self))
+            # However, the line below relies on the stack which may not be the
+            # safest current choice.
```

Comment:
@b-butler Could you elaborate on those cases? This implementation appears more fragile to me.